crates/astria-conductor/src/executor/channel.rs

```diff
@@ -0,0 +1,194 @@
+        Self
+    }
+}
+
+impl<T> From<TokioSendError<T>> for SendError {
+    fn from(_: TokioSendError<T>) -> Self {
+        Self
+    }
+}
+
+// allow: this is mimicking tokio's `SendError` that returns the stack-allocated object.
+#[allow(clippy::result_large_err)]
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub(crate) enum TrySendError<T> {
+    #[error("the channel is closed")]
+    Closed(T),
+    #[error("no permits available")]
+    NoPermits(T),
+}
+
+impl<T> TrySendError<T> {
+    fn from_semaphore(err: &TryAcquireError, block: T) -> Self {
+        match err {
+            tokio::sync::TryAcquireError::Closed => Self::Closed(block),
+            tokio::sync::TryAcquireError::NoPermits => Self::NoPermits(block),
+        }
+    }
+}
+
+impl<T> From<TokioSendError<T>> for TrySendError<T> {
+    fn from(err: TokioSendError<T>) -> Self {
+        Self::Closed(err.0)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct Sender<T> {
+    sem: Arc<Semaphore>,
+    chan: UnboundedSender<T>,
+}
+
+impl<T> Sender<T> {
+    /// Sends a block, waiting until the channel has permits.
+    ///
+    /// Returns an error if the channel is closed.
+    pub(super) async fn send(&self, block: T) -> Result<(), SendError> {
+        let permit = self.sem.acquire().await?;
+        permit.forget();
+        self.chan.send(block)?;
+        Ok(())
+    }
+
+    /// Attempts to send a block without blocking.
+    ///
+    /// Returns an error if the channel is out of permits or if it has been closed.
+    // allow: this is mimicking tokio's `SendError` that returns the stack-allocated object.
+    #[allow(clippy::result_large_err)]
+    pub(super) fn try_send(&self, block: T) -> Result<(), TrySendError<T>> {
+        let permit = match self.sem.try_acquire() {
+            Ok(permit) => permit,
+            Err(err) => return Err(TrySendError::from_semaphore(&err, block)),
+        };
+        permit.forget();
+        self.chan.send(block)?;
+        Ok(())
+    }
+}
+
+pub(super) struct Receiver<T> {
+    cap: usize,
+    sem: Arc<Semaphore>,
+    chan: UnboundedReceiver<T>,
+}
+
+impl<T> Receiver<T> {
+    /// Sets the channel's capacity to `cap`.
+    ///
+    /// `cap` will be the maximum number of blocks that can be sent
+    /// over the channel before new permits are added with `[SoftBlockReceiver::add_permits]`.
+    pub(super) fn set_capacity(&mut self, cap: usize) {
+        self.cap = cap;
+    }
+
+    /// Adds up to `capacity` number of permits to the channel.
+    ///
+    /// `capacity` is previously set by [`SoftBlockReceiver::set_capacity`]
+    /// or zero by default.
+    pub(super) fn fill_permits(&self) {
+        let additional = self.cap.saturating_sub(self.sem.available_permits());
+        self.sem.add_permits(additional);
+    }
+
+    /// Receives a block over the channel.
+    pub(super) async fn recv(&mut self) -> Option<T> {
+        self.chan.recv().await
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
This test sweet has all tests which expect `TrySendError::NoPermits(())`. I'd like to have a test for closed as well as a success path.

If I were to replace `try_send` with just return an `TrySendError::NoPermits(())` wouldn't these all pass?

Reply:
This was a good call. Adding tests to check for the channel being closed revealed that the `Closed` variant was never hit because the channel senders held an `Arc`. I have downgraded the senders to hold a `Weak` instead, and am also explicitly closing the semaphore when dropping the receiver. I have also added tests for the immediate `try_send`, the async `send`, and the `recv` methods.